src/python/pants/help/help_info_extracter.py

```diff
@@ -810,27 +838,43 @@ def _extract_api_types() -> Iterator[tuple[type, str, tuple[type, ...]]]:
             ),
         )
 
-        def get_api_type_info_loader(api_type: type) -> Callable[[], PluginAPITypeInfo]:
+        def get_api_type_info(api_types: list[type]):
+            """Gather the info from each of the types and aggregate it.
+
+            The gathering is the expensive operation, and we can only aggregate once we've gathered.
+            """
+
             def load() -> PluginAPITypeInfo:
-                return PluginAPITypeInfo.create(
-                    api_type,
-                    rules,
-                    provider=", ".join(type_graph[api_type]["providers"]),
-                    dependencies=type_graph[api_type]["dependencies"],
-                    dependents=type_graph[api_type].get("dependents", ()),
-                    union_members=tuple(
-                        sorted(member.__name__ for member in union_membership.get(api_type))
-                    ),
-                )
+                gatherered_infos = [
+                    PluginAPITypeInfo.create(
+                        api_type,
+                        rules,
+                        provider=type_graph[api_type]["providers"],
+                        dependencies=type_graph[api_type]["dependencies"],
+                        dependents=type_graph[api_type].get("dependents", ()),
+                        union_members=tuple(
+                            sorted(member.__qualname__ for member in union_membership.get(api_type))
+                        ),
+                    )
+                    for api_type in api_types
+                ]
+                return reduce(lambda x, y: x.merged_with(y), gatherered_infos)
 
             return load
 
-        return LazyFrozenDict(
-            {
-                f"{api_type.__module__}.{api_type.__name__}": get_api_type_info_loader(api_type)
-                for api_type in sorted(all_types, key=attrgetter("__name__"))
-            }
-        )
+        # We want to provide a lazy dict so we don't spend so long doing the info gathering.
+        # We provide a list of the types here, and the lookup function performs the gather and the aggregation
+        names_to_types: dict[str, list[type]] = defaultdict(list)
+        for api_type in sorted(all_types, key=attrgetter("__qualname__")):
+            names_to_types[PluginAPITypeInfo.fully_qualified_name_from_type(api_type)].append(
+                api_type
+            )
+
+        infos: dict[str, Callable[[], PluginAPITypeInfo]] = {
+            k: get_api_type_info(v) for k, v in names_to_types.items()
+        }
```

Comment:
This could be written using a `groupby`:
```suggestion
        api_type_name=PluginAPITypeInfo.fully_qualified_name_from_type
        all_types_sorted=sorted(all_types, key=api_type_name)
        infos: dict[str, Callable[[], PluginAPITypeInfo]] = {
            k: get_api_type_info(tuple(v)) 
            for k, v in itertools.groupby(all_types_sorted, key=api_type_name)
        }
```